python/cuml/svm/svm.pyx

```diff
@@ -0,0 +1,297 @@
+        # The pointer is stored with type size_t (see fit()), because we
+        # cannot have cdef CppSVC[float, float] *svc here.
+        #
+        # Alternatively we could have a thin cdef class PyCppSVC wrapper around
+        # CppSVC, or we could make SVC itself a cdef class.
+        #
+        # CppSVC is not yet created here because the data type will be only
+        # known when we call fit(). Similarily kernel_params is not yet created
+        # here, because gamma can depend on the input data.
+
+    def __dealloc__(self):
+        # deallocate CppSVC
+        cdef CppSVC[float]* svc_f
+        cdef CppSVC[double]* svc_d
+        cdef KernelParams *kernel_params
+        if self.svcHandle is not None:
+            if self.dtype == np.float32:
+                svc_f = <CppSVC[float]*><size_t> self.svcHandle
+                del svc_f
+            elif self.dtype == np.float64:
+                svc_d = <CppSVC[double]*><size_t> self.svcHandle
+                del svc_d
+            else:
+                raise TypeError("Unknown type for SVC class")
+        if self.kernel_params is not None:
+            kernel_params = <KernelParams*><size_t> self.kernel_params
+            del kernel_params
+
+    def _get_c_kernel(self, kernel):
+        return {
+            'linear': LINEAR,
+            'poly': POLYNOMIAL,
+            'rbf': RBF,
+            'sigmoid': TANH
+        }[kernel]
+
+    def _gamma_val(self, X):
+        # Calculate the value for gamma kernel parameter
+        if type(self.gamma) is str:
+            if self.gamma == 'auto':
+                return 1 / self.n_cols
+            elif self.gamma == 'scale':
+                x_std = X.std()
+                return 1 / (self.n_cols * x_std)
+            else:
+                raise ValueError("Not implemented gamma option: " + self.gamma)
+        else:
+            return self.gamma
+
+    def fit(self, X, y):
+        """
+        Fit the model with X and y.
+
+        Parameters
+        ----------
+        X : array-like (device or host) shape = (n_samples, n_features)
+            Dense matrix (floats or doubles) of shape (n_samples, n_features).
+            Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+            ndarray, cuda array interface compliant array like CuPy
+
+        y : array-like (device or host) shape = (n_samples, 1)
+            Dense vector (floats or doubles) of shape (n_samples, 1).
+            Acceptable formats: cuDF Series, NumPy ndarray, Numba device
+            ndarray, cuda array interface compliant array like CuPy
+
+        """
+
+        cdef uintptr_t X_ptr, y_ptr
+
+        X_m, X_ptr, self.n_rows, self.n_cols, self.dtype = \
+            input_to_dev_array(X, order='F')
+
+        y_m, y_ptr, _, _, _ = input_to_dev_array(y)
+
+        cdef CppSVC[float]* svc_f = NULL
+        cdef CppSVC[double]* svc_d = NULL
+        cdef KernelParams *kernel_params
```

Comment:
Will need to take special care during pickling to make sure this is written out and read back in properly.

Reply:
The KernelParams and svmParameter structures are constructed on the fly from the attributes of the SVC object. Only one cdef structure is kept in memory, the svmModel. This could also be constructed on the fly, because all its members are expressed as attributes of the SVC object. Pickling takes care of restoring the svmModel struct.